extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCacheImpl.java

```diff
@@ -293,4 +301,38 @@ private <T> T cast(Object value) {
                     "An existing cached value type does not match the type returned by the value loading function", e);
         }
     }
+
+    @SuppressWarnings("unchecked")
+    private <K, V> Uni<V> computeWithStats(K key, Function<K, Uni<V>> valueLoader) {
+        return (Uni<V>) Uni.createFrom()
+                .completionStage(cache.asMap().compute(key,
+                        new BiFunction<Object, CompletableFuture<Object>, CompletableFuture<Object>>() {
+                            @Override
+                            public CompletableFuture<Object> apply(Object key, CompletableFuture<Object> value) {
+                                if (value == null) {
+                                    statsCounter.recordMisses(1);
+                                    return (CompletableFuture<Object>) valueLoader.apply((K) key)
+                                            .map(i -> NullValueConverter.toCacheValue(i))
+                                            .subscribeAsCompletionStage();
+                                } else {
+                                    LOGGER.tracef("Key [%s] found in cache [%s]", key, cacheInfo.name);
+                                    statsCounter.recordHits(1);
+                                    return value;
+                                }
+                            }
+                        }));
+    }
+
+    @SuppressWarnings("unchecked")
+    private <K, V> Uni<V> computeWithoutStats(K key, Function<K, Uni<V>> valueLoader) {
+        return (Uni<V>) Uni.createFrom()
+                .completionStage(cache.asMap().computeIfAbsent(key, new Function<Object, CompletableFuture<Object>>() {
```

Comment:
`computeIfAbsent` won't work well with the `UnresolvedUniValue` logic from `CacheResultInterceptor`:
https://github.com/quarkusio/quarkus/blob/4585831c288124f78a8fdc8065dc3e56bf8972eb/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheResultInterceptor.java#L66-L82

The cache may contain an `UnresolvedUniValue.INSTANCE` value associated with a given key.

Reply:
Hm, I need to take a look at the `UnresolvedUniValue.INSTANCE` and what it's used for...

Reply:
When `@CacheResult` is used on a method returning `Uni`, the `Uni` itself is never cached but the item it emits is eventually cached after the `Uni` has been resolved. Between the `Uni` subscription and its resolution, if the cache does not contain any existing value associated with the key, then `UnresolvedUniValue.INSTANCE` is put into the cache. When the item is eventually emitted, `CacheResultInterceptor` replaces the value from the cache (which can be `UnresolvedUniValue.INSTANCE` or an actual value) with the emitted item.

Reply:
But `CaffeineCacheImpl.get(K, Function<K, V>)` does not reflect the `UnresolvedUniValue.INSTANCE` either, or am I missing something?

Reply:
By the way, I don't think that it's a common use case to combine `@CacheResult` and `Cache#get()`. OTOH I agree that the behavior should be consistent.

Reply:
Hm, from the code in `CacheResultInterceptor` it looks like it might happen that parallel `@CacheResult` invocations may result in multiple cache writes :thinking:.

Reply:
> But `CaffeineCacheImpl.get(K, Function<K, V>)` does not reflect the `UnresolvedUniValue.INSTANCE` either, or am I missing something?

That correct, but `CaffeineCacheImpl.get(K, Function<K, V>)` does not deal with async value loaders so `UnresolvedUniValue.INSTANCE` isn't really important in that method.

Reply:
> By the way, I don't think that it's a common use case to combine @CacheResult and Cache#get(). OTOH I agree that the behavior should be consistent.

I agree.

Reply:
> Hm, from the code in `CacheResultInterceptor` it looks like it might happen that parallel `@CacheResult` invocations may result in multiple cache writes 🤔.

Correct. That was done on purpose and is mentioned in a comment:
https://github.com/quarkusio/quarkus/blob/a2777c31e5b3978c2f10404c7762c76a1d00ce10/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCacheImpl.java#L220-L224

This was done because of the lazy nature of `Uni`s which can lead to multiple writes.

Reply:
The most recent write will always win in case of parallel `Uni` resolutions for the same cache key.

Reply:
> > But `CaffeineCacheImpl.get(K, Function<K, V>)` does not reflect the `UnresolvedUniValue.INSTANCE` either, or am I missing something?
> 
> That correct, but `CaffeineCacheImpl.get(K, Function<K, V>)` does not deal with async value loaders so `UnresolvedUniValue.INSTANCE` isn't really important in that method.

I don't understand. If a method annotated with `@CacheResult` and `Cache#get()` are called in parallel for the same key then `Cache#get()` may obtain the `UnresolvedUniValue.INSTANCE`, or?

Reply:
Yes, that could happen. Maybe we should do something about it.